File: CPE/Exam/240521/pE_UVA11518.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

vector<vector<int>> g;
vector<bool> visited;

int dfs(int u) {
    visited[u] = true;
    int res = 1;
    for (int v : g[u]) {
        if (!visited[v]) {
            res += dfs(v);
        }
    }
    return res;
}
int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int t, n, m, l, u, v, z;
    cin >> t;
    while (t--) {
        cin >> n >> m >> l;
        g.clear();
        g.resize(n);
        for (int i=0; i<m; i++) {
            cin >> u >> v;
            g[u-1].push_back(v-1);
        }
        int ans = 0;
        visited.assign(n, false);
        for (int i=0; i<l; i++) {
            cin >> z;
            if (!visited[z-1]) {
                ans += dfs(z-1);
            }
        }
        cout << ans << endl;
    }
    return 0;
}

File: CPE/Exam/240521/pE_UVA11518.py
"""
    DFS
    是問 l 次推倒後，總共有幾個骨牌會被推倒
    
    AC: UVA, CPE, ZeroJudge
"""
t = int(input())

def dfs(u):
    visited[u] = True
    res = 1 # u 倒下
    for v in g[u]:
        if not visited[v]: # v 未倒下
            res += dfs(v)
    return res

for _ in range(t):
    n, m, l = map(int, input().split())
    g = [[] for _ in range(n)]
    
    for _ in range(m):
        u, v = map(int, input().split())
        g[u-1].append(v-1)

    ans = 0
    visited = [False] * n
    for _ in range(l):
        z = int(input())
        if not visited[z-1]:
            ans += dfs(z-1)
    print(ans)